### PR TITLE
Login api implementation pt. 1

### PR DESCRIPTION
**Pull request by vexonius**
DI is omitted (because of PR size) so this is only base networking implementation with login endpoint.
Interceptor pattern is used for the learning purposes and global config instead adding header on every request manually. It isn't pretty, I know

### CODE REVIEW

File: QuizApp/Data/Common/Helpers/URLSession+Extension.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+                    return continuation.resume(throwing: error)
+                }
+
+                continuation.resume(returning: (data, response))
+            }
+
+            task.resume()
+        }
+    }
```
Comment (marinzg) [inline QuizApp/Data/Common/Helpers/URLSession+Extension.swift]: Why do you need this?

File: QuizApp/Data/Common/Networking/Api.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+enum Api {
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/Api.swift]: Why is this an enum if you're only using `static let`? Can't it be a `struct`?

File: QuizApp/Data/Common/Networking/BaseNetworkClient.swift [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        params: [String: String],
+        headers: [String: String],
+        requestBody: T
+    ) async throws -> U
+
+}
+
+class BaseNetworkClient {
+
+    static let baseApiUrl = Api.apiURL
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/BaseNetworkClient.swift]: Why is it static and are you using it anywhere?

File: QuizApp/Data/Common/Networking/BaseNetworkClient.swift [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+extension BaseNetworkClient: BaseNetworkClientProtocol {
+
+    func get<T: Codable>(
+        url: URL,
+        params: [String: String],
+        headers: [String: String],
+        responseType: T.Type
+    ) async throws -> T {
+
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/BaseNetworkClient.swift]: There's no need for a first empty line in functions.

File: QuizApp/Data/Common/Networking/BaseNetworkClient.swift [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private func encodeRequestBody<T: Encodable>(model: T) async throws -> Data {
+        return try encoder.encode(model)
+    }
+
+    private func decodeResponseBody<T: Decodable>(type: T.Type, data: Data) async throws -> T {
+        return try decoder.decode(T.self, from: data)
+    }
+
+    private func validateResponse(response: URLResponse) async throws {
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/BaseNetworkClient.swift]: Why is this async?

File: QuizApp/Data/Common/Networking/BaseNetworkClient.swift [hunk 6]
```diff
@@ -0,0 +1,183 @@
+import Foundation
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/BaseNetworkClient.swift]: I would reorganize this file a bit:

1. HttpMethod, NetworkError, NetworkResponseStatus and ClientError should be in a separate file (for example `BaseClientModels`
2. BaseNetworkClientProtocol also in a separate file.
3. I wouldn't implement `BaseNetworkClientProtocol` in an extension, but rather in the original implementation and I'd put it on top. That way when you open `BaseNetworkClient` first functions you'd see are those that are called from the outside and the rest of the logic is "hidden" lower in the file.

Also, you're using `async` and `await` in places where it's not necessary, like:
```swift
private func encodeRequestBody<T: Encodable>(model: T) throws -> Data {
    try encoder.encode(model)
}
```
where the signature of the `encode` function is:
```swift
open func encode<T>(_ value: T) throws -> Data where T : Encodable
```
which in no way indicates that it's async.

And base client logic should go to `QuizApp/Core` folder, while the exact clients should be in `QuizApp/Feature/Client` or `QuizApp/Client/Feature/` which ever you prefer.

File: QuizApp/Data/Common/Networking/BaseNetworkClient.swift [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    private var urlSession: URLSession = {
+        var config = URLSessionConfiguration.default
+        config.protocolClasses = [Interceptor.self]
+
+        return URLSession(configuration: config)
+    }()
+
+    private lazy var encoder: JSONEncoder = {
+        return JSONEncoder()
```
Comment (marinzg) [inline QuizApp/Data/Common/Networking/BaseNetworkClient.swift]: We're not using `return` in one line computed properties and functions.

File: QuizApp/Data/Login/Network/LoginNetworkClient.swift [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+    func login(requestBody: LoginRequestBodyNetworkModel) async throws -> LoginResponseNetworkModel {
+        try await baseNetworkClient.post(
+            url: LoginEndpoints.login.path,
+            params: [:],
+            headers: [HeaderField.contentType.key: HeaderValue.defaultContentType.value],
+            requestBody: requestBody)
+    }
+
+}
```
Comment (marinzg) [inline QuizApp/Data/Login/Network/LoginNetworkClient.swift]: This also doesn't have to be extracted to an extension, but should be in a separate file.

**Review by marinzg — CHANGES_REQUESTED** [review]
In the future, if you're implementing a layer from scratch, please create one PR for each layer.

File: QuizApp/Data/Common/Helpers/URLSession+Extension.swift (hunk 1 above)
Comment (vexonius) [reply QuizApp/Data/Common/Helpers/URLSession+Extension.swift]: async await for URLSession.dataTask is only available in iOS 15, this backports it

**Review by marinzg — APPROVED** [review]

**Pull request merged**